### ISSUE TITLE
Can't compile: senml struct issues

### ISSUE THREAD

**Issue by bsmelo** [i2ot/i2ot#1]
After going through the installation of the dependencies pointed out in the `README.md` file, I've run `make` and, it seems that there's a problem with the senml struct usage.

Any idea on how to solve this?

```
[  2:12PM ]  [ bruno@lain:~/coap-apps/i2ot(master✔) ]
 $ make
mkdir -p bin; \
        mkdir -p obj; \
        cd src; \
        make
make[1]: Entering directory `/home/bruno/coap-apps/i2ot/src'
test -d "microcoap" || git clone --depth=1 https://github.com/i2ot/microcoap.git; \
        make all OBJDIR=/home/bruno/coap-apps/i2ot/src/../obj/ CFLAGS+=-DENDPOINT_DYNAMIC  -C microcoap
Cloning into 'microcoap'...
remote: Counting objects: 10, done.
remote: Compressing objects: 100% (10/10), done.
remote: Total 10 (delta 0), reused 4 (delta 0), pack-reused 0
Unpacking objects: 100% (10/10), done.
Checking connectivity... done.
make[2]: Entering directory `/home/bruno/coap-apps/i2ot/src/microcoap'
gcc -DENDPOINT_DYNAMIC -std=c99 -Wall -Wextra -Werror -O2 -c coap.c -o /home/bruno/coap-apps/i2ot/src/../obj/coap.o
make[2]: Leaving directory `/home/bruno/coap-apps/i2ot/src/microcoap'
test -d "senml" || git clone --depth=1 https://github.com/i2ot/senml.git; \
        make OBJDIR=/home/bruno/coap-apps/i2ot/src/../obj/ -C senml
Cloning into 'senml'...
remote: Counting objects: 7, done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 7 (delta 0), reused 4 (delta 0), pack-reused 0
Unpacking objects: 100% (7/7), done.
Checking connectivity... done.
make[2]: Entering directory `/home/bruno/coap-apps/i2ot/src/senml'
gcc -std=gnu99 -Wall -Wextra -Werror -O2 -c senml.c -o /home/bruno/coap-apps/i2ot/src/../obj/senml.o
make[2]: Leaving directory `/home/bruno/coap-apps/i2ot/src/senml'
gcc -std=gnu99 -Wall -Wextra -Werror -O2 -c services.c -o /home/bruno/coap-apps/i2ot/src/../obj/services.o
services.c: In function ‘thingspeak_us’:
services.c:419:52: error: ‘senml_base_info_t’ has no member named ‘base_value_f’
                         float_val = pack->base_info->base_value_f;
                                                    ^
services.c:424:53: error: ‘senml_base_info_t’ has no member named ‘base_value_s’
                         string_val = pack->base_info->base_value_s;
                                                     ^
services.c:427:86: error: ‘senml_record_t’ has no member named ‘value_f’
                         sprintf(&request[strlen(request)], "%f", float_val + curr_rec->value_f);
                                                                                      ^
services.c:430:41: error: ‘senml_record_t’ has no member named ‘value_s’
                                 curr_rec->value_s);
                                         ^
services.c:433:41: error: ‘senml_record_t’ has no member named ‘value_b’
                                 curr_rec->value_b ? "true" : "false");
                                         ^
make[1]: *** [services.o] Error 1
make[1]: Leaving directory `/home/bruno/coap-apps/i2ot/src'
make: *** [all] Error 2
```

**Comment by backenklee** [comment]
Hi Bruno,
thanks for reporting this! Originally, the unions were unnamed, but that only works with `--std=gnu99` and not with `--std=c99`. Hence I named the unions, but forgot to update the code that accessed their members. I have committed a fix just now; if you pull from the master tree, it should build.

**Comment by backenklee** [comment]
Side note: this implementation of SenML really needs some love, the [implementation for RIOT](https://github.com/RIOT-OS/RIOT/pull/5544) is much more up to date.

**Comment by bsmelo** [comment]
Hi @backenklee,
Thanks for the prompt answer!

I was able to make it now, but I can't run it:
```
[ 10:03AM ]  [ bruno@lain:~/coap-apps/i2ot_2/bin(master✔) ]
 $ ./i2ot 
ERROR: service not available: (null)
[ 10:03AM ]  [ bruno@lain:~/coap-apps/i2ot_2/bin(master✔) ]
 $ ./i2ot --help
Usage: ./i2ot [options]
Options:
        -c, --config FILE       Specify the location of the config file
        -i, --interface IF      Specify the interface i2ot should use for communication with clients
[ 10:03AM ]  [ bruno@lain:~/coap-apps/i2ot_2/bin(master✔) ]
 $ ./i2ot -c ../i2ot.conf 
../i2ot.conf:1: no such option 'gateway'
```

Note that I'm using the configuration file that comes from GitHub (https://github.com/i2ot/i2ot/blob/master/i2ot.conf).

**Comment by backenklee** [comment]
Hi @bsmelo ,

again, my bad. i²ot needs to know what to do with the data it receives, and originally, users could specify different actions for different devices. I dropped this feature later, so all you have to specify is the service to use (e.g. [thingspeak.com](https://www.thingspeak.com)), authentication credentials, and (if needed) optional attributes.

I have [fixed](https://github.com/i2ot/i2ot/commit/9cf2b73f91da19cd2ec38b648039a5883a81d637) the default config and added explanations as comments inside it, I hope that helps. You can a list of available services at the very bottom of the [services.c](https://github.com/i2ot/i2ot/blob/master/src/services.c) file. At the moment, only thingspeak works.

**Comment by bsmelo** [comment]
Hi @backenklee,

Again thanks a lot for your help. I believe I'm getting closer.

I can successfully run the program now:
```
[  3:06PM ]  [ bruno@lain:~/coap-apps/i2ot/bin(master✗) ]
 $ ./i2ot -c ../i2ot.conf
```
I've created a new Data Channel on Thingspeak with one field, called "dummySensor" and changed the config file to hold my `api_key = "THINGSPEAK_WRITE_KEY"`.

I believe what I should do now is to send a CoAP POST Request with payload `dummySensor=5` to i2ot, so it can send this data to my channel on Thingspeak, right? Problem is, I tried sending this CoAP POST Request using Firefox Copper (to `coap://localhost:5683/`) and it's not working. I'm not sure what the problem is, but it seems the i2ot program doesn't have any ports open to receive the requests:
```
[  3:09PM ]  [ bruno@lain:~/Dropbox/coap-apps/i2ot/bin(master✗) ]
 $ netstat -tu -p | grep i2ot
(Not all processes could be identified, non-owned process info
 will not be shown, you would have to be root to see it all.)
```

Do you have any idea what can be wrong, or what I might be doing wrong?
Also, thanks [again] for your help and support.

**Comment by backenklee** [comment]
Hi @bsmelo ,

I tried your setup, and the quick fix seems to be using 127.0.0.1 instead of localhost (couldn't figure out why, though). Other problems I've encountered so far are: 

- IPv6 not enabled (especially if you're using a Raspberry Pi or Debian)
- Firewall interfering (even on localhost)

Also, netstat doesn't show programs that are just listening by default; add -l to see them (then i2ot shows up, too). While you're still experimenting, I recommend to build i2ot with `make debug` to see what's happening. By default, it uses IPv4, but by adding `-UIPV4` to the compiler flags (or removing the `#define` in main.c) you can force it to use IPv6. Let me know if there any other issues!

Until then, happy hacking :-)

**Comment by backenklee** [comment]
@bsmelo sorry, forgot a few things yesterday: you don't have to create a channel yourself; instead, put your account API key in the config. When a device sends a POST request to /i2ot/rs, i2ot will create a new channel for the device via the thingspeak API.

Second, Wireshark seems to capture packets before the firewall drops them (at least on my openSUSE system with iptables).

And finally, there's a more detailed description of how i2ot works in my [BA thesis](https://github.com/backenklee/bachelor_thesis/blob/master/thesis.pdf). There's also pseudo code of the core, the modules (for different platforms), and a client in the appendix. That should give you an overview of the structure and functioning.

I'm going to close this issue now since the original problem has been resolved; feel free to open another one if anything else comes up!
